File: wazero/testdata/c/guest.c
#include <emscripten.h>
#include <stdint.h>
#include <stdlib.h>

/* READ ZIG SOURCE CODE TO UNDERSTAND THIS FILE */

// Reseved memory for input, which mmap will be mapped to,
// in some area of reserved_input space.
uint8_t* reserved_input_memory;
uint32_t reserved_input_size;

// Memory that will be used for input, which is a part of reserved_input_memory,
// and it's mmaped.
uint8_t* input_memory;
uint32_t input_size;

__attribute__((export_name("_start")))
void _start() {}

extern uint64_t  karmap_mem_ack(uint32_t ptr);

__attribute__((export_name("karmap_mem_req")))
EMSCRIPTEN_KEEPALIVE uint32_t karmap_mem_req(uint32_t size) {
    if (reserved_input_memory == NULL) {
        reserved_input_memory = malloc(size);
        reserved_input_size = size;
    }
    
    uint64_t offsetsize = karmap_mem_ack((uint32_t)reserved_input_memory);
    
    input_memory = reserved_input_memory;
    input_memory += offsetsize >> 32;
    input_size = offsetsize & 0xFFFFFFFF;
    
    return input_size;
}

__attribute__((export_name("work")))
EMSCRIPTEN_KEEPALIVE uint32_t work() {
    uint32_t i = 0;
    uint32_t acc = 0;

    uint32_t * input_memory_as_u32 = (uint32_t*)input_memory;
    uint32_t input_size_as_u32 = input_size / 4;
    while(i < input_size_as_u32) {
        acc += input_memory_as_u32[i];
        i++;
    }
    return acc;
}

// Only used for debugging/testing/bechmarking
__attribute__((export_name("debug_test_malloc")))
EMSCRIPTEN_KEEPALIVE uint32_t debug_test_malloc(uint32_t size) {
    input_memory = malloc(size);
    return (uint32_t)input_memory;
}

// Only used for debugging/testing/bechmarking
__attribute__((export_name("debug_try_write")))
EMSCRIPTEN_KEEPALIVE void debug_try_write() {
    input_memory[0] = 0xFF;
    input_memory[input_size - 1] = 42;
}

File: wazero/testdata/zig/guest.zig
const std = @import("std");

// reserved_input_memory is used to "reserve" memory for input_memory,
// in some GC-languages, that is mandatory to avoid been lost by GC.
//
// The "reserved" is larger than input_memory, because mmap requires
// the size of the memory to be a multiple of the page size, so
// reserved_input_memory can be 128KB larger than required. The
// mmap will happen inside reserved_input_memory.
//
// In Zig, it is not necessary, but it is a good practice to do so,
// to make clear that such memory is reserved for input_memory.
var reserved_input_memory: []u8 = undefined;

// input_memory is the memory that is used to store the input data.
//
// It is a slice of reserved_input_memory, which is mmaped.
var input_memory: []u8 = undefined;

pub fn main() void {}

// karmap_mem_ack is called by the guest, passing the pointer to the
// reserved_input_memory. The host will then mmap the memory region
// (or part of it) and inform the guest about the offset and size,
// in that order, packed in a u64.
extern "env" fn karmap_mem_ack(ptr: u32) u64;

// karmap_mem_req is called by the host, forcing the guest to create a memory
// region that will be used to store the input data.
//
// Then, the guest will call karmap_mem_ack, passing the pointer to the
// host. The host will then mmap the memory region (or part of it) and
// inform the guest about the size and offset (based on reserved_input_memory).
export fn karmap_mem_req(size: u32) u32 {
    if (reserved_input_memory.len == 0) {
        reserved_input_memory = std.heap.page_allocator.alloc(u8, size) catch unreachable;
    }

    var offsetsize: u64 = karmap_mem_ack(@ptrToInt(reserved_input_memory.ptr));

    input_memory = reserved_input_memory[@intCast(usize, offsetsize >> 32)..@intCast(usize, ((offsetsize >> 32) + (offsetsize & 0xFFFFFFFF)))];

    return @intCast(u32, offsetsize);
}

export fn work() u32 {
    var acc: u32 = 0;

    var slice: [2]usize = [2]usize{@ptrToInt(input_memory.ptr), input_memory.len / 4};
    var slice_32: []u32 = @ptrCast(*[]u32, &slice).*;
    for (slice_32) |value| {
        acc += value;
    }

    return acc;
}

// Only used for debugging/testing/bechmarking
export fn debug_test_malloc(size: u32) u32 {
    input_memory = std.heap.page_allocator.alloc(u8, size) catch unreachable;
    return @ptrToInt(input_memory.ptr);
}

// Only used for debugging/testing/bechmarking
export fn debug_try_write() void {
    input_memory[0] = 0xFF;
    input_memory[input_memory.len - 1] = 42;
}